cohortextractor/tpp_backend.py

```diff
@@ -2787,6 +2788,180 @@ def patients_outpatient_appointment_date(
+        # Allow for case insensitive matching and whitespace
+        therapeutic_matches = to_list(with_these_therapeutics)
+        if therapeutic_matches:
+            fragments = [
+                f"Intervention LIKE {pattern} ESCAPE '!'"
+                for pattern in codelist_to_like_patterns(
+                    therapeutic_matches, prefix="%", suffix="%"
+                )
+            ]
+            filter_conditions.append("(" + " OR ".join(fragments) + ")")
+
+        indication_matches = to_list(with_these_indications)
+        if indication_matches:
+            valid_indications = [
+                "hospital_onset",
+                "hospitalised_with",
+                "non_hospitalised",
+            ]
+            for indication in indication_matches:
+                assert (
+                    indication in valid_indications
+                ), f"'{indication}' is not a valid indication; options are {', '.join(valid_indications)}"
+            indications = [quote(indication) for indication in indication_matches]
+            filter_conditions.append(f"COVID_indication IN ({', '.join(indications)})")
+
+        date_condition, date_joins = self.get_date_condition(
+            "t", "TreatmentStartDate", between
+        )
+        filter_conditions.append(date_condition)
+
+        where_filter_conditions = (
+            f"WHERE {' AND '.join(filter_conditions)}" if filter_conditions else ""
+        )
+
+        # Result ordering
+        if find_first_match_in_period:
+            ordering = "ASC"
+        else:
+            ordering = "DESC"
+
+        use_partition_query = True
+        if returning == "binary_flag" or returning == "date":
+            returning = "binary_flag"
+            column_definition = "1"
+        elif returning == "therapeutic":
+            # remove whitespace, convert to comma-separated string
+            column_definition = "REPLACE(LTRIM(RTRIM(Intervention)), ' and ', ',')"
+        elif returning == "risk_group":
+            # First remove any "Patients with a" and replace " and " with "," within individual risk group fields
+            # Then join the 3 risk cohort fields with ","
+            # Note that the last "s" in SOT02_risk_cohorts is correct
+            parts = [
+                f"REPLACE(REPLACE({risk_col}, 'Patients with a ', ''), ' and ', ',')"
+                for risk_col in [
+                    "MOL1_high_risk_cohort",
+                    "SOT02_risk_cohorts",
+                    "CASIM05_risk_cohort",
+                ]
+            ]
+            # coalesce the parts with a leading ','
+            coalesced_parts = " + ".join(
+                f"coalesce(',' + NULLIF({part}, ''), '')" for part in parts
+            )
+            # use stuff to remove the first ','
+            column_definition = f"STUFF({coalesced_parts}, 1, 1, '')"
```

Comment:
There was a request to also remove duplicates that may appear in this joined string. That's fairly simple if TPP was using a modern version of SQL server, but since we can't use STRING_SPLIT, it woud involve a complicated XML parse which doesn't seem worth it. Especially given that removing duplicates should be pretty simple after data extraction, and also currently I don't think there are any patients with risk groups in more than one of these fields, so it's likely to be an infrequent occurrence anyway